dapps/src/api/time.rs

```diff
@@ -56,27 +64,72 @@ impl From<io::Error> for Error {
 	fn from(err: io::Error) -> Self { Error::Io(format!("{}", err)) }
 }
 
-impl From<fetch::Error> for Error {
-	fn from(err: fetch::Error) -> Self { Error::Fetch(format!("{}", err)) }
+impl From<ntp::errors::Error> for Error {
+	fn from(err: ntp::errors::Error) -> Self { Error::Ntp(format!("{}", err)) }
 }
 
-
 /// Time provider.
 pub trait TimeProvider: Clone + Send + 'static {
 	/// Returns an instance of this provider.
 	fn new() -> Self where Self: Sized;
-	/// Returns current time in milliseconds.
-	fn utc_timestamp_millis(&self) -> i64;
+	/// Returns current time.
+	fn now(&self) -> Timespec;
 }
 /// Default system time provider.
 #[derive(Clone)]
 pub struct StdTimeProvider;
 impl TimeProvider for StdTimeProvider {
 	fn new() -> Self where Self: Sized { StdTimeProvider }
-	fn utc_timestamp_millis(&self) -> i64 {
-		let time = ::time::now_utc().to_timespec();
+	fn now(&self) -> Timespec {
+		::time::now_utc().to_timespec()
+	}
+}
+
+/// NTP time drift checker.
+pub trait Ntp {
+	/// Returns the current time drift.
+	fn drift(&self) -> BoxFuture<Duration, Error>;
+}
 
-		1_000 * time.sec + time.nsec as i64 / 1_000_000
+/// NTP client using the SNTP algorithm for calculating drift.
+#[derive(Clone)]
+pub struct SimpleNtp<T> {
+	address: Arc<String>,
+	time_provider: T,
+	pool: CpuPool,
```

Comment:
One more grumble. Could we share this `CpuPool` with `fetch`? Otherwise we'll spawn 8 threads and most of them will do nothing 99% of the time.